Shopify/cli
# Implement 'app dev' support for Functions via draft versions

Implements `app dev` support for function extensions by watching for file changes, automatically rebuilding, and pushing an extension draft.

### WHY are these changes introduced?

To align Functions with other extensions and enable a better inner loop DX.

### WHAT is this pull request doing?

* Performs an initial build and draft deploy of all function extensions on `dev` start.
* Enables a file watcher which builds and draft deploys functions on any changes.
* Adds a `watch` setting to function extensions to allow customizing the file watcher paths.

These new behaviors only apply for unified/simplified deployments.

### How to test your changes?

1. Create a new app
1. Add a function extension
1. Enable unified/simplified deployments for the app
1. If using a non-JavaScript/TypeScript template, add the watch configuration to your function toml, e.g.
    ```
    [build]
    command = "cargo wasi build --release"
    path = "target/wasm32-wasi/release/order-discount-rust.wasm"
    watch = "src/**/*.rs"
    ```
1. Run `shopify app dev` on the app, adding `--verbose` to validate draft deployments
1. Verify build and deploy on launch and when you make file changes.

### Example output (end of startup + rust file change)
```
2023-06-22 16:27:41 │ extensions   │ Draft updated successfully for extension: order-discount-javascript
2023-06-22 16:27:41 │ extensions   │ Draft updated successfully for extension: order-discount-typescript
2023-06-22 16:27:41 │ extensions   │ Draft updated successfully for extension: checkout-validation-js
2023-06-22 16:27:42 │ extensions   │ Draft updated successfully for extension: order-discount-assemblyscript
2023-06-22 16:27:44 │ extensions   │ Draft updated successfully for extension: order-discount-cpp
2023-06-22 16:27:45 │ extensions   │    Compiling order-discount-rust v1.0.0 (/Users/nickwesselman/src/polyglot-functions/app/extensions/order-discount-rust)
2023-06-22 16:27:46 │ extensions   │ wasm-opt -Oz --zero-filled-memory --strip-producers -o build/order-discount-golang.wasm build/order-discount-golang-unopt.wasm
2023-06-22 16:27:48 │ extensions   │ Draft updated successfully for extension: order-discount-golang
2023-06-22 16:27:48 │ extensions   │     Finished release [optimized] target(s) in 14.24s
2023-06-22 16:27:48 │ extensions   │   Optimizing with wasm-opt
2023-06-22 16:27:50 │ extensions   │ Draft updated successfully for extension: order-discount-rust
2023-06-22 16:27:50 │ extensions   │ Function extension order-discount-cpp is missing the 'build.watch' setting, automatic builds are disabled.
2023-06-22 16:27:50 │ extensions   │ Function extension order-discount-assemblyscript is missing the 'build.watch' setting, automatic builds are disabled.
2023-06-22 16:27:50 │ extensions   │ Function extension order-discount-golang is missing the 'build.watch' setting, automatic builds are disabled.
2023-06-22 16:27:50 │ extensions   │ Function extension order-discount-zig is missing the 'build.watch' setting, automatic builds are disabled.
2023-06-22 16:29:36 │ extensions   │ Config file at path /Users/nickwesselman/src/polyglot-functions/app/extensions/order-discount-rust/shopify.function.extension.toml changed
2023-06-22 16:29:37 │ extensions   │ Draft updated successfully for extension: order-discount-rust
2023-06-22 16:29:41 │ extensions   │ Config file at path /Users/nickwesselman/src/polyglot-functions/app/extensions/order-discount-rust/shopify.function.extension.toml changed
2023-06-22 16:29:42 │ extensions   │ Draft updated successfully for extension: order-discount-rust
```
